### DISCUSSION TITLE
Why does context.request_repaint() from timer thread not refresh the UI?

### DISCUSSION THREAD

**Discussion by nnmm** [emilk/egui#2128] [Q&A] [answered]
I have a simple `eframe 0.19` app:

```rust
use eframe::egui::{self, Widget};

struct MyApp;

impl MyApp {
    fn new(cc: &eframe::CreationContext<'_>) -> Self {
        let egui_context = cc.egui_ctx.clone();
        std::thread::spawn(move || {
            loop {
                std::thread::sleep(std::time::Duration::from_millis(100));
                egui_context.request_repaint();
            }
        });
        Self
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.label(format!("{:?}", std::time::Instant::now()));
        });
    }
}

fn main() {
    eframe::run_native(
        "myapp",
        eframe::NativeOptions::default(),
        Box::new(move |cc| Box::new(MyApp)),
    );
}
```

I would expect this to continually refresh the timestamp, but instead the app only refreshes on mouse movement. What is wrong with this?

**Comment by emilk** [comment]
I agree that this should work, and I don't know why it doesn't for you. What platform are you on?

I thin you'll need to dig into this yourself (I don't have the time right now). Here are some places to start:

Calling `egui_context.request_repaint` should call this:
https://github.com/emilk/egui/blob/367378d75dcc1670418ce464cb1aafa4c8b2387b/crates/eframe/src/native/run.rs#L391-L393

This should send a `RequestRepaintEvent` to the event loop, which is caught here:

https://github.com/emilk/egui/blob/367378d75dcc1670418ce464cb1aafa4c8b2387b/crates/eframe/src/native/run.rs#L103-L106

Which a few lines later triggers `window.request_redraw`: https://github.com/emilk/egui/blob/367378d75dcc1670418ce464cb1aafa4c8b2387b/crates/eframe/src/native/run.rs#L142

which finally should trigger either `Event::RedrawEventsCleared` or `Event::RedrawRequested` at: https://github.com/emilk/egui/blob/367378d75dcc1670418ce464cb1aafa4c8b2387b/crates/eframe/src/native/run.rs#L91-L101

If you add a few calls to `tracing::info!` in these places (and run with `RUST_LOG=debug`) you should have enough info to figure out where things break down.

Good luck, and keep us posted!

**Reply by nnmm** [reply]
My fault, sorry! In `main()`, I forgot to update `Box::new(MyApp)` to `Box::new(MyApp::new(cc))`.

**Comment by nnmm** [comment, marked as answer]
See https://github.com/emilk/egui/discussions/2128#discussioncomment-3839874
